tommath.h

```diff
@@ -567,6 +567,69 @@ int mp_prime_next_prime(mp_int *a, int t, int bbs_style);
  */
 int mp_prime_random_ex(mp_int *a, int t, int size, int flags, ltm_prime_callback cb, void *dat);
 
+/*
+ * All MP_xBIT sizes need one data type that has twice the size of "x",
+ * that means that the type uint16_t must be available, even for MP_8BIT.
+ */
+#ifdef MP_8BIT
+#   define LTM_SIEVE_BIGGEST_PRIME      65521lu
+#   define LTM_SIEVE_UINT               uint16_t
+#   define LTM_SIEVE_PR_UINT            PRIu16
+#   define LTM_SIEVE_UINT_MAX           0xFFFFlu
+#   define LTM_SIEVE_UINT_MAX_SQRT      0xFFlu
+#elif ( (defined MP_64BIT) && (defined LTM_SIEVE_USE_LARGE_SIEVE) )
+#   define LTM_SIEVE_BIGGEST_PRIME      18446744073709551557llu
+#   define LTM_SIEVE_UINT               uint64_t
+#   define LTM_SIEVE_PR_UINT            PRIu64
+#   define LTM_SIEVE_UINT_MAX           0xFFFFFFFFFFFFFFFFllu
+#   define LTM_SIEVE_UINT_MAX_SQRT      0xFFFFFFFFllu
+#else
+#   define LTM_SIEVE_BIGGEST_PRIME      4294967291lu
+#   define LTM_SIEVE_UINT               uint32_t
+#   define LTM_SIEVE_PR_UINT            PRIu32
+#   define LTM_SIEVE_UINT_MAX           0xFFFFFFFFlu
+#   define LTM_SIEVE_UINT_MAX_SQRT      0xFFFFlu
+#endif
```

Comment:
Could you move those definitions to the private header with a MP_* prefix? Maybe just use size_t instead of LTM_SIEVE_UINT?

Reply:
> Could you move those definitions to the private header with a MP_* prefix

It already needs rebase'ing, so: no problem.
(Apropos rebase'ing: should I wait a day or two until your stuff gets merged or is there much more to come?)

> Maybe just use size_t instead of LTM_SIEVE_UINT

Mmh&hellip;no, I need to know the exact sizes here and `size_t` is most likely `LTM_SIEVE_UINT` but I cannot be completely sure and I need to.

Reply:
@czurnieden From me there is not much more to come for now. I would rather like to reduce the backlog a bit. 

> Mmh…no, I need to know the exact sizes here and size_t is most likely LTM_SIEVE_UINT but I cannot be completely sure and I need to.

We already include limits.h?

Reply:
(Sorry for the delay, everybody seems to have pushed their urgent things to "after Easter" which didn't make it better \*sigh\*)

> Could you move those definitions to the private header with a MP_* prefix?

The `MP_` prefix I can do. Was already planned when I watched you harmonizing the style.

Making (some of) the macros private, not so much.

The sieve exists in three sizes, one for 8-bit, one extra-large, and one for the whole rest.

```
#ifdef MP_8BIT
#   define LTM_SIEVE_BIGGEST_PRIME      65521lu
#   define LTM_SIEVE_UINT               uint16_t
#   define LTM_SIEVE_PR_UINT            PRIu16
#   define LTM_SIEVE_UINT_MAX           0xFFFFlu
#   define LTM_SIEVE_UINT_MAX_SQRT      0xFFlu
#elif ( (defined MP_64BIT) && (defined LTM_SIEVE_USE_LARGE_SIEVE) )
#   define LTM_SIEVE_BIGGEST_PRIME      18446744073709551557llu
#   define LTM_SIEVE_UINT               uint64_t
#   define LTM_SIEVE_PR_UINT            PRIu64
#   define LTM_SIEVE_UINT_MAX           0xFFFFFFFFFFFFFFFFllu
#   define LTM_SIEVE_UINT_MAX_SQRT      0xFFFFFFFFllu
#else
#   define LTM_SIEVE_BIGGEST_PRIME      4294967291lu
#   define LTM_SIEVE_UINT               uint32_t
#   define LTM_SIEVE_PR_UINT            PRIu32
#   define LTM_SIEVE_UINT_MAX           0xFFFFFFFFlu
#   define LTM_SIEVE_UINT_MAX_SQRT      0xFFFFlu
#endif
```

- The macro `LTM_SIEVE_UINT` is used in the definition of the functions, a replacement is not easy:

>  Maybe just use size_t instead of LTM_SIEVE_UINT?

According to the standard (ISO/IEC 9899:2011 7.20.3) the limit of `size_t` is `SIZE_MAX` and has only a minimum (65535) not a fixed size. I would need to use a large type (`mp_word` would do it, I think) for it as a replacement and that would make it slower for high-mp architectures and is also a waste of memory.

I could also do something in the line of `typedef TYPE mp_small_prime` which looks a bit more elegant.

- `LTM_SIEVE_BIGGEST_PRIME` can be set private, although it comes quite handy in loops.
- `LTM_SIEVE_PR_UINT` can be set private, although it would be a bit of a work for the user to find out the correct way to print the values.
- `LTM_SIEVE_UINT_MAX` can be set private. It is convenient but `LTM_SIEVE_BIGGEST_PRIME` should suffice.
- `LTM_SIEVE_UINT_MAX_SQRT`  can be set private, but it is the size of the base which might be usefull to know for low memory architectures, On the other side: it is not directly the size, it needs a bit of work to get the actual size from this information. This is the macro I would be able to kick out with the cleanest conscious.

Tell me what you think while I'll run the rebase and the `replace LTM_SIEVE_ MP_SIEVE_  -- *.[ch]`